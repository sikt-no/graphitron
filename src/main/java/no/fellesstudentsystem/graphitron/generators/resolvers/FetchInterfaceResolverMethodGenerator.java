package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.definitions.objects.AbstractObjectDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.InterfaceDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverMethodGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.Dependency;
import no.fellesstudentsystem.graphitron.generators.dependencies.QueryDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;
import no.fellesstudentsystem.graphitron.generators.abstractions.DBClassGenerator;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphitron.generators.db.FetchDBClassGenerator.SAVE_DIRECTORY_NAME;
import static no.fellesstudentsystem.graphitron.mappings.JavaPoetClassName.*;
import static org.apache.commons.lang3.StringUtils.capitalize;
import static org.apache.commons.lang3.StringUtils.uncapitalize;

/**
 * Generates resolvers for queries returning an interface. E.g. the node resolver.
 */
public class FetchInterfaceResolverMethodGenerator extends ResolverMethodGenerator<ObjectField> {
    private final ObjectDefinition localObject;

    public FetchInterfaceResolverMethodGenerator(ObjectDefinition localObject, ProcessedSchema processedSchema) {
        super(localObject, processedSchema);
        this.localObject = localObject;
    }

    @Override
    public MethodSpec generate(ObjectField target) {
        InterfaceDefinition interfaceDefinition = processedSchema.getInterface(target);
        TypeName returnClassName = interfaceDefinition.getGraphClassName();

        var spec = getDefaultSpecBuilder(target.getName(), returnClassName);

        if (!localObject.isRoot()) {
            spec.addParameter(localObject.getGraphClassName(), uncapitalize(localObject.getName()));
        }

        var inputField = target.getInputFields().get(0);
        String inputFieldName = inputField.getName();
        spec
                .addParameter(inputIterableWrap(inputField), inputFieldName)
                .addParameter(DATA_FETCHING_ENVIRONMENT.className, ENV_NAME)
                .addStatement("String tablePartOfId = $T.getTablePartOf($N)", FIELD_HELPERS.className, inputFieldName)
                .addCode("\n");

        processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(it -> it.implementsInterface(interfaceDefinition.getName()))
                .sorted(Comparator.comparing(AbstractObjectDefinition::getName))
                .forEach(implementation -> {
                    var queryLocation = implementation.getName() + DBClassGenerator.FILE_NAME_SUFFIX;
                    var map = Map.ofEntries(
                            Map.entry("env", ENV_NAME),
                            Map.entry("string", STRING.className),
                            Map.entry("returnType", returnClassName),
                            Map.entry("dataloader", DATA_LOADER_FACTORY.className),
                            Map.entry("batchLoader", MAPPED_BATCH_LOADER_WITH_CONTEXT.className),
                            Map.entry("implementationClass", implementation.getGraphClassName()),
                            Map.entry("future", COMPLETABLE_FUTURE.className),
                            Map.entry("queryInstanceField", uncapitalize(queryLocation)),
                            Map.entry("implName", implementation.getName()),
                            Map.entry("inputFieldName", inputFieldName),
                            Map.entry("inputFieldNameCap", capitalize(inputFieldName)),
                            Map.entry("selectionSets", SELECTION_SET.className),
                            Map.entry("environmentUtils", ENVIRONMENT_UTILS.className),
                            Map.entry("referenceFieldName", capitalize(target.getName())),
                            Map.entry("context", Dependency.CONTEXT_NAME)
                    );
                    dependencySet.add(new QueryDependency(queryLocation, SAVE_DIRECTORY_NAME));

                    spec.beginControlFlow("if (tablePartOfId.equals($T.$N.getViewId().toString()))", TABLES.className, implementation.getTable().getMappingName());
                    spec.addStatement(CodeBlock.builder().addNamed(
                                    "return $env:N.getDataLoaderRegistry().<$string:T, $returnType:T>computeIfAbsent(tablePartOfId, name ->$W" +
                                            "$dataloader:T.newMappedDataLoader(($batchLoader:T<$string:T, $implementationClass:T>) (keys, loaderEnvironment) ->$>$W" +
                                            "$future:T.completedFuture($queryInstanceField:N.load$implName:NBy$inputFieldNameCap:NsAs$referenceFieldName:N" +
                                            "($context:N, keys, new $selectionSets:T($environmentUtils:T.getSelectionSetsFromEnvironment(loaderEnvironment))))))" +
                                            "$<$W.load($inputFieldName:N, $env:N)", map)
                            .build());
                    spec.endControlFlow();
                });

        spec.addStatement("throw new $T(\"could not find dataloader for $N with prefix \" + tablePartOfId)", ILLEGAL_ARGUMENT_EXCEPTION.className, inputFieldName);

        return spec.build();
    }

    @Override
    public List<MethodSpec> generateAll() {
        return localObject
                .getFields()
                .stream()
                .filter(ObjectField::isGenerated)
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .map(this::generate)
                .collect(Collectors.toList());
    }

    @Override
    public boolean generatesAll() {
        return localObject
                .getFields()
                .stream()
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .allMatch(ObjectField::isGenerated);
    }
}
