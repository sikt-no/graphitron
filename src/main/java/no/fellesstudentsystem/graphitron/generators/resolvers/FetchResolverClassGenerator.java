package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.TypeSpec;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverClassGenerator;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphql.mapping.GraphQLReservedName.SCHEMA_ROOT_NODE_MUTATION;

/**
 * Class generator for basic select resolver classes.
 */
public class FetchResolverClassGenerator extends ResolverClassGenerator<ObjectDefinition> {
    public final String SAVE_DIRECTORY_NAME = ResolverClassGenerator.DEFAULT_SAVE_DIRECTORY_NAME + ".query";

    public FetchResolverClassGenerator(ProcessedSchema processedSchema) {
        super(processedSchema);
    }

    @Override
    public void generateQualifyingObjectsToDirectory(String path, String packagePath) throws IOException {
        var classes = processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(ObjectDefinition::isGenerated)
                .filter(obj -> !obj.getName().equals(SCHEMA_ROOT_NODE_MUTATION.getName()))
                .map(this::generate)
                .collect(Collectors.toList());

        for (var generatedClass : classes) {
            writeToFile(generatedClass, path, packagePath, SAVE_DIRECTORY_NAME);
        }
    }

    @Override
    public TypeSpec generate(ObjectDefinition target) {
        return getSpec(
                target.getName(),
                List.of(
                        new FetchResolverMethodGenerator(target, processedSchema),
                        new FetchInterfaceResolverMethodGenerator(target, processedSchema)
                )
        ).build();
    }

    @Override
    public String getDefaultSaveDirectoryName() {
        return SAVE_DIRECTORY_NAME;
    }
}
