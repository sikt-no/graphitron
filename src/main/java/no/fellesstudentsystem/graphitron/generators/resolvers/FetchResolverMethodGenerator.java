package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterizedTypeName;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverMethodGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.Dependency;
import no.fellesstudentsystem.graphitron.generators.dependencies.QueryDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;
import no.fellesstudentsystem.graphql.naming.GraphQLReservedName;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.Map.entry;
import static no.fellesstudentsystem.graphitron.generators.abstractions.DBClassGenerator.FILE_NAME_SUFFIX;
import static no.fellesstudentsystem.graphitron.generators.codebuilding.FormatCodeBlocks.returnCompletedFuture;
import static no.fellesstudentsystem.graphitron.generators.context.ClassNameFormat.wrapListIf;
import static no.fellesstudentsystem.graphitron.generators.context.NameFormat.asQueryMethodName;
import static no.fellesstudentsystem.graphitron.generators.db.FetchCountDBMethodGenerator.TOTAL_COUNT_NAME;
import static no.fellesstudentsystem.graphitron.generators.db.FetchDBClassGenerator.SAVE_DIRECTORY_NAME;
import static no.fellesstudentsystem.graphitron.mappings.JavaPoetClassName.*;
import static org.apache.commons.lang3.StringUtils.capitalize;
import static org.apache.commons.lang3.StringUtils.uncapitalize;

/**
 * This class generates the resolvers for default fetch queries with potential arguments or pagination.
 */
public class FetchResolverMethodGenerator extends ResolverMethodGenerator<ObjectField> {
    private static final String
            EXECUTION_STEP_PATH_ID_DELIMITER = "||",
            BATCHED_ENV_NAME = "batchEnvLoader",
            PAGE_SIZE_NAME = "pageSize",
            AFTER_NAME = "after",
            MAP_RESULT_NAME = "mapResult",
            QUERY_RESULT_NAME = "dbResult",
            PAGINATION_RESULT_NAME = "pagedResult",
            RESULT_ENTRY_NAME = "resultEntry",
            RESULT_VALUE_NAME = "resultValue",
            PAGINATION_RESULT_ENTRY_NAME = "pagedResultEntry",
            SELECTION_SET_NAME = "selectionSet";
    private static final int MAX_NUMBER_OF_NODES_TO_DISPLAY = 1000;

    public FetchResolverMethodGenerator(ObjectDefinition localObject, ProcessedSchema processedSchema) {
        super(localObject, processedSchema);
    }

    @Override
    public MethodSpec generate(ObjectField target) {
        var localObject = getLocalObject();
        var isRootLevel = localObject.isRoot();

        var returnClassName = getReturnTypeName(target);
        var spec = getDefaultSpecBuilder(target.getName(), returnClassName);

        if (!isRootLevel) {
            spec.addParameter(localObject.getGraphClassName(), uncapitalize(localObject.getName()));
        }

        var allQueryInputs = getQueryInputs(spec, target);
        spec.addParameter(DATA_FETCHING_ENVIRONMENT.className, ENV_NAME);

        spec.addCode(queryMethodCalls(target, returnClassName, allQueryInputs));

        var currentResultName = (isRootLevel) ? QUERY_RESULT_NAME : MAP_RESULT_NAME;
        if (target.hasForwardPagination()) {
            if (!isRootLevel) {
                spec
                        .addCode("var $L = $N.entrySet().stream().map(", PAGINATION_RESULT_NAME, currentResultName)
                        .beginControlFlow("$L ->", RESULT_ENTRY_NAME)
                        .addStatement("var $L = $N.getValue()", RESULT_VALUE_NAME, RESULT_ENTRY_NAME);
                currentResultName = RESULT_VALUE_NAME;
            }
            var nodeType = processedSchema.getConnectionObject(target).getNodeType();
            var paginationInputMap = Map.ofEntries(
                    entry("math", MATH.className),
                    entry("dbResult", currentResultName),
                    entry("resultName", (!isRootLevel) ? PAGINATION_RESULT_ENTRY_NAME : PAGINATION_RESULT_NAME),
                    entry("pageSize", PAGE_SIZE_NAME),
                    entry("connectionClass", RELAY_CONNECTION_IMPL.className),
                    entry("connectionCursorClass", RELAY_CONNECTION_CURSOR_IMPL.className),
                    entry("pageInfoClass", RELAY_PAGE_INFO_IMPL.className),
                    entry("nodeType", nodeType),
                    entry("edgeClass", RELAY_EDGE_IMPL.className),
                    entry("collectors", COLLECTORS.className),
                    entry("integerClass", INTEGER.className),
                    entry("totalCount", TOTAL_COUNT_NAME),
                    entry("maxNoOfNodes", MAX_NUMBER_OF_NODES_TO_DISPLAY)
            );
            spec.addNamedCode(getPaginationSegmentString(), paginationInputMap);
            currentResultName = PAGINATION_RESULT_NAME;
            if (!isRootLevel) {
                spec
                        .addStatement(
                                "return new $T<$T, $T<$N>>($N.getKey(), $N)",
                                SIMPLE_ENTRY.className,
                                STRING.className,
                                RELAY_CONNECTION.className,
                                nodeType,
                                RESULT_ENTRY_NAME,
                                PAGINATION_RESULT_ENTRY_NAME
                        )
                        .endControlFlow(").collect($T.toMap(r -> r.getKey(), r -> r.getValue()))", COLLECTORS.className);
            }
        } // TODO: Backwards pagination if necessary.

        return spec
                .addCode(returnCompletedFuture(currentResultName))
                .addCode(getMethodCallTail(target))
                .build();
    }

    @NotNull
    private ArrayList<String> getQueryInputs(MethodSpec.Builder spec, ObjectField referenceField) {
        var allQueryInputs = new ArrayList<String>();
        for (var input : referenceField.getNonReservedInputFields()) {
            var name = input.getName();
            spec.addParameter(inputIterableWrap(input), name);
            allQueryInputs.add(name);
        }
        if (referenceField.hasForwardPagination()) {
            spec
                    .addParameter(INTEGER.className, GraphQLReservedName.PAGINATION_FIRST.getName())
                    .addStatement(
                            "int " + PAGE_SIZE_NAME + " = $T.ofNullable($N).orElse(" + referenceField.getFirstDefault() + ")",
                            OPTIONAL.className,
                            GraphQLReservedName.PAGINATION_FIRST.getName()
                    );
            allQueryInputs.add(PAGE_SIZE_NAME);
            spec.addParameter(STRING.className, GraphQLReservedName.PAGINATION_AFTER.getName());
            allQueryInputs.add(GraphQLReservedName.PAGINATION_AFTER.getName());
        }
        return allQueryInputs;
    }

    private TypeName getReturnTypeName(ObjectField referenceField) {
        var refClassName = processedSchema.getObject(referenceField).getGraphClassName();
        var hasFPagination = referenceField.hasForwardPagination();
        TypeName returnClassName = wrapListIf(
                refClassName,
                referenceField.isIterableWrapped() || getLocalObject().isRoot() && !hasFPagination
        );
        if (hasFPagination) {
            var connectionObject = processedSchema.getConnectionObject(referenceField);
            TypeName nodeClassName = processedSchema.getObject(connectionObject.getNodeType()).getGraphClassName();
            returnClassName = ParameterizedTypeName.get(RELAY_CONNECTION.className, nodeClassName);
        }
        return returnClassName;
    }

    @NotNull
    private CodeBlock queryMethodCalls(ObjectField referenceField, TypeName returnClassName, ArrayList<String> allQueryInputs) {
        var localObject = getLocalObject();

        var queryMethodName = asQueryMethodName(referenceField.getName(), localObject.getName());
        var dbQueryCallCodeBlock = CodeBlock.builder();

        var queryLocation = localObject.getName() + FILE_NAME_SUFFIX;

        var selectionSetDeclaration = CodeBlock.of("var $L = new $T($T.getSelectionSetsFromEnvironment($N))",
                SELECTION_SET_NAME,
                referenceField.hasForwardPagination() ? CONNECTION_SELECTION_SET.className : SELECTION_SET.className,
                ENVIRONMENT_UTILS.className,
                localObject.isRoot() ? ENV_NAME : BATCHED_ENV_NAME);
        dbQueryCallCodeBlock.addStatement(selectionSetDeclaration);

        dbQueryCallCodeBlock.add("var $L = $N.$L(", QUERY_RESULT_NAME, uncapitalize(queryLocation), queryMethodName);
        dependencySet.add(new QueryDependency(queryLocation, SAVE_DIRECTORY_NAME));

        allQueryInputs.add(SELECTION_SET_NAME);

        if (!localObject.isRoot()) {
            allQueryInputs.add(0, "idSet");
        }

        if (!allQueryInputs.isEmpty()) {
            dbQueryCallCodeBlock.addStatement("$N, $L)", Dependency.CONTEXT_NAME, String.join(", ", allQueryInputs));
        }

        if (referenceField.hasRequiredPaginationFields()) {
            dbQueryCallCodeBlock.addStatement("var $L = selectionSet.contains($S) ? $N.count$L($N, $L) : null",
                    TOTAL_COUNT_NAME,
                    TOTAL_COUNT_NAME,
                    uncapitalize(queryLocation),
                    capitalize(queryMethodName),
                    Dependency.CONTEXT_NAME,
                    allQueryInputs.stream()
                            .filter(it -> !it.equals(PAGE_SIZE_NAME) && !it.equals(AFTER_NAME) && !it.equals(SELECTION_SET_NAME))
                            .collect(Collectors.joining(", "))
            );
        }

        var methodBodyBuilder = CodeBlock.builder();

        if (localObject.isRoot()) {
            methodBodyBuilder.add(dbQueryCallCodeBlock.build());
        } else {
            methodBodyBuilder
                    .beginControlFlow(
                            "$T<$T, $T> loader = $N.getDataLoaderRegistry().computeIfAbsent(\""
                                    + queryMethodName
                                    + "\""
                                    + ", name ->",
                            DATA_LOADER.className,
                            STRING.className,
                            returnClassName,
                            ENV_NAME
                    )
                    .beginControlFlow("var batchLoader = ($T<$T, $T>) (keys, " + BATCHED_ENV_NAME + ") ->",
                            MAPPED_BATCH_LOADER_WITH_CONTEXT.className, STRING.className, returnClassName)
                    .addStatement("var keyToId = keys.stream().collect($W$T.toMap(s -> s, s -> s.substring(s.lastIndexOf($S) + $L)))",
                            COLLECTORS.className, EXECUTION_STEP_PATH_ID_DELIMITER, EXECUTION_STEP_PATH_ID_DELIMITER.length())
                    .addStatement("var idSet = new $T<>(keyToId.values())", HASH_SET.className)
                    .add(dbQueryCallCodeBlock.build())
                    .addStatement(
                            "var "
                                    + MAP_RESULT_NAME
                                    + " = keyToId.entrySet().stream()$W"
                                    + ".filter(it -> $N.get(it.getValue()) != null)$W"
                                    + ".collect($T.toMap($T.Entry::getKey, it -> $N.get(it.getValue())))",
                            QUERY_RESULT_NAME,
                            COLLECTORS.className,
                            MAP.className,
                            QUERY_RESULT_NAME
                    );
        }
        return methodBodyBuilder.build();
    }

    @NotNull
    private CodeBlock getMethodCallTail(ObjectField referenceField) {
        var localObject = getLocalObject();
        var closeMethodCall = CodeBlock.builder();
        if (!localObject.isRoot()) {
            closeMethodCall
                    .endControlFlow("")
                    .addStatement("return $T.newMappedDataLoader(batchLoader)", DATA_LOADER_FACTORY.className)
                    .endControlFlow(")")
                    .add(
                            "return loader.load($N.getExecutionStepInfo().getPath().toString() + $S + $N.getId(), $N)",
                            ENV_NAME,
                            EXECUTION_STEP_PATH_ID_DELIMITER,
                            uncapitalize(localObject.getName()),
                            ENV_NAME
                    );

            if (referenceField.getFieldType().isIterableNonNullable()) {
                closeMethodCall.addStatement(".thenApply(data -> $T.ofNullable(data).orElse($T.of()))", OPTIONAL.className, LIST.className);
            } else {
                closeMethodCall.add(";");
            }
        }
        return closeMethodCall.build();
    }

    /**
     * Template-style code for the large pagination section.
     */
    private String getPaginationSegmentString() {
        return "var size = $math:T.min($dbResult:N.size(), $pageSize:N);\n" +
                "var items = $dbResult:N.subList(0, size);\n" +
                "var firstItem = items.size() == 0 ? null : new $connectionCursorClass:T(items.get(0).getId());\n" +
                "var lastItem = items.size() == 0 ? null : new $connectionCursorClass:T(items.get(items.size() - 1).getId());\n" +
                "var $resultName:L = $connectionClass:T\n" +
                "        .<$nodeType:N>builder()\n" +
                "        .setPageInfo(\n" +
                "                new $pageInfoClass:T(firstItem, lastItem, false, $dbResult:N.size() > $pageSize:N)\n" +
                "        )\n" +
                "        .setNodes(items)\n" +
                "        .setEdges(\n" +
                "                items\n" +
                "                        .stream()\n" +
                "                        .map(item -> new $edgeClass:T<$nodeType:N>(item, new $connectionCursorClass:T(item.getId())))\n" +
                "                        .collect($collectors:T.toList())\n" +
                "        )\n" +
                "        .setTotalCount($totalCount:N != null ? $math:T.min($maxNoOfNodes:L, $totalCount:N) : null)\n" +
                "        .build();\n";
    }

    @Override
    public List<MethodSpec> generateAll() {
        return getLocalObject()
                .getReferredFieldsFromObjectNames(processedSchema.getNamesWithTableOrConnections())
                .stream()
                .filter(ObjectField::isGenerated)
                .map(this::generate)
                .collect(Collectors.toList());
    }

    @Override
    public boolean generatesAll() {
        var fieldStream = getLocalObject().getReferredFieldsFromObjectNames(processedSchema.getNamesWithTableOrConnections()).stream();
        return getLocalObject().isRoot()
                ? fieldStream.allMatch(ObjectField::isGenerated)
                : fieldStream.allMatch(f -> !f.isResolver() || f.isGenerated());
    }
}
