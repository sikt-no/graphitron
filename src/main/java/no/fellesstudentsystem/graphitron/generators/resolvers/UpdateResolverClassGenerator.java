package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.TypeSpec;
import no.fellesstudentsystem.graphitron.definitions.fields.AbstractField;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.generators.abstractions.MethodGenerator;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverClassGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.ServiceDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;

import java.util.List;
import java.util.stream.Collectors;

import static org.apache.commons.lang3.StringUtils.capitalize;

/**
 * Class generator for basic update resolver classes.
 */
public class UpdateResolverClassGenerator extends ResolverClassGenerator<ObjectField> {
    public static final String
            INTERFACE_FILE_NAME_SUFFIX = "MutationResolver",
            SAVE_DIRECTORY_NAME = "mutation";

    public UpdateResolverClassGenerator(ProcessedSchema processedSchema) {
        super(processedSchema);
    }

    @Override
    public void generateQualifyingObjectsToDirectory(String path, String packagePath) {
        var mutation = processedSchema.getMutationType();
        if (mutation != null && mutation.isGenerated()) {
            var classes = mutation
                    .getFields()
                    .stream()
                    .filter(ObjectField::isGenerated)
                    .map(this::generate)
                    .collect(Collectors.toList());

            for (var generatedClass : classes) {
                writeToFile(generatedClass, path, packagePath, getDefaultSaveDirectoryName());
            }
        }
    }

    protected void setDependencies(List<MethodGenerator<? extends AbstractField<?>>> generators, TypeSpec.Builder spec) {
        generators
                .stream()
                .flatMap(gen -> gen.getDependencySet().stream())
                .distinct()
                .sorted()
                .filter(dep -> !(dep instanceof ServiceDependency)) // Inelegant solution, but it should work for now.
                .forEach(dep -> spec.addField(dep.getSpec()));
    }

    @Override
    public TypeSpec generate(ObjectField target) {
        return getSpec(
                capitalize(target.getName()),
                List.of(
                        new ServiceUpdateResolverMethodGenerator(target, processedSchema),
                        new MutationTypeResolverMethodGenerator(target, processedSchema)
                )
        ).build();
    }

    @Override
    public String getDefaultSaveDirectoryName() {
        return ResolverClassGenerator.DEFAULT_SAVE_DIRECTORY_NAME + "." + SAVE_DIRECTORY_NAME;
    }

    @Override
    public String getExpectedInterfaceSuffix() {
        return INTERFACE_FILE_NAME_SUFFIX;
    }
}
