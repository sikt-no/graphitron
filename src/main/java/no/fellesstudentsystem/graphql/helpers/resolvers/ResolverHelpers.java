package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;

import java.util.Optional;

public class ResolverHelpers {
    public static int getPageSize(Integer first, int max, int defaultMax) {
        return Optional.ofNullable(first).map(it -> Math.min(max, it)).orElse(defaultMax);
    }

    public static DSLContext selectContext(DataFetchingEnvironment env, DSLContext defaultContext) {
        return env.getLocalContext() == null ? defaultContext : (DSLContext) env.getLocalContext();
    }

    @NotNull
    public static SelectionSet getSelectionSet(DataFetchingEnvironment env) {
        return new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(env));
    }
}
