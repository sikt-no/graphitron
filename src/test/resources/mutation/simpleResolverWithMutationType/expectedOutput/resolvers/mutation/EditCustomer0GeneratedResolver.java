package fake.code.generated.resolvers.mutation;

import fake.code.generated.queries.mutation.EditCustomer0DBQueries;
import fake.code.generated.transform.RecordTransformer;
import fake.graphql.example.api.EditCustomer0MutationResolver;
import fake.graphql.example.model.EditInput;
import fake.graphql.example.model.EditResponse0;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Override;
import java.lang.String;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.resolvers.ResolverHelpers;
import org.jooq.DSLContext;

public class EditCustomer0GeneratedResolver implements EditCustomer0MutationResolver {
    @Inject
    DSLContext ctx;

    @Override
    public CompletableFuture<EditResponse0> editCustomer0(String id, EditInput in,
            DataFetchingEnvironment env) throws Exception {
        var ctx = ResolverHelpers.selectContext(env, this.ctx);

        var transform = new RecordTransformer(env, this.ctx);

        var inRecord = transform.editInputToJOOQRecord(in, "in");

        var rowsUpdated = EditCustomer0DBQueries.editCustomer0(ctx, id, inRecord);

        var editResponse0 = new EditResponse0();
        editResponse0.setId0(id);

        return CompletableFuture.completedFuture(editResponse0);
    }
}
