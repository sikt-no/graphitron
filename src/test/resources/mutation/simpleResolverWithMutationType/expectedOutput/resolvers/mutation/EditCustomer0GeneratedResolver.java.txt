package fake.code.generated.resolvers.mutation;

import fake.code.generated.queries.mutation.EditCustomer0DBQueries;
import fake.graphql.example.package.api.EditCustomer0MutationResolver;
import fake.graphql.example.package.model.EditInput;
import fake.graphql.example.package.model.EditResponse0;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Override;
import java.lang.String;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.arguments.Arguments;
import no.sikt.graphitron.jooq.generated.testdata.tables.records.CustomerRecord;
import org.jooq.DSLContext;

public class EditCustomer0GeneratedResolver implements EditCustomer0MutationResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private EditCustomer0DBQueries editCustomer0DBQueries;

    @Override
    public CompletableFuture<EditResponse0> editCustomer0(String id, EditInput in,
            DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        var flatArguments = Arguments.flattenArgumentKeys(env.getArguments());

        var inRecord = new CustomerRecord();
        inRecord.attach(ctx.configuration());

        if (in != null) {
            if (flatArguments.contains("in/firstName")) {
                inRecord.setFirstName(in.getFirstName());
            }
        }

        var rowsUpdated = editCustomer0DBQueries.editCustomer0(ctx, id, inRecord);

        var editResponse0 = new EditResponse0();
        editResponse0.setId0(id);

        return CompletableFuture.completedFuture(editResponse0);
    }
}